### PR TITLE
HTTP cache control

### LINKED ISSUE #650: Update Kiwix-serve cache strategy
Kiwix serve delivers many types of content:
1.  ZIM content
2. Kiwix-serve static content
3. Kiwix-serve dynamic content (API, search, etc.)

For each type, here is how we want to cache:
1. ZIM content: cache as long as the (concerned) content does not change
2. Kiwix-serve static content: cache as long as we don't run a different version of kiwix-serve
3. Kiwix-serve dynamic content: cache as long as the (concerned) library has not been updated

Which implies 3 different strategies...

## ZIM content

ZIM content does give us no freedom about the URLs calls and we have no way to garanty that what is available at the URL at two different times is the same content. Therefore the content can be cached but must be revalidate (with `Etag` issued from the ZIM ID). To avoid a systematic revalidation we can set a max-age of one hour like `Cache-Control: max-age=3600, must-revalidate`.

## Kiwix-serve static content

This is the easier case, we need to "cache busting", for example by adding `?UNIQUE_ID` at the end of the URL of any of the resources. This `UNIQUE_ID` should be fixed at compilation time. `Cache-Control: max-age=31536000, immutable` then should be enough (so no other related HTTP headers, like  `Etag`, `Last-Modified` or `Expires`).

## Kiwix-serve dynamic content

The dynamic content is consistent and can be cached as long as the library is not updated or the server restarted. I would propose to force systematically revalidation with an `Etag` which is changed each time the internal library is changed. The cache policy would look like `Cache-Control: max-age=0, must-revalidate`.

### LINKED ISSUE #831: Cache control of static resources
This is a sub-ticket of #650

#712 has introduced support for versioning static resources, however the HTTP caching side of their handling has not been addressed.

Currently a build of `kiwix-serve` containing static resource `/skin/RESOURCE` with the cache id value `CACHEID` is supposed only to receive and handle requests of the form `/skin/RESOURCE?cacheid=CACHEID`. #650 proposes responding to such requests with `Cache-Control: max-age=31536000, immutable`. That's easy.

Questions:

1. What should be the `Cache-Control` response to a request *without any* cache-id (`/skin/RESOURCE`)?
    a. #716 has introduced a special user-facing static resource served at `/viewer` which is going to be accessed without any cache id. What should be the caching policy for that one (especially in interaction with kiwix/kiwix-tools#17)?
2. What should be the response to a request *with a wrong* cache-id (`/skin/RESOURCE?cacheid=WRONGCACHEID`)?

### PR DESCRIPTION
**Pull request by veloman-yunkan**
Fixes #650
Fixes #831

### CODE REVIEW

**Comment by kelson42** [comment]
Fixes https://github.com/kiwix/libkiwix/issues/831 and partly fixes https://github.com/kiwix/libkiwix/issues/650?! Might be better.

**Comment by veloman-yunkan** [comment]
> Fixes #831 and partly fixes #650?! Might be better.

No. This PR in its current state fixes #831 only partly since it doesn't address the aspects discussed in the questions posed in that ticket. Since the higher level ticket #650 was not concerned with such issues, it can be considered to be fully fixed by this PR.

**Comment by veloman-yunkan** [comment]
Now #831 is fully fixed by this PR

File: src/server/internalServer.cpp [hunk 1]
```diff
@@ -1033,6 +1035,11 @@ std::unique_ptr<Response> InternalServer::handle_content(const RequestContext& r
            + suggestSearchMsg(searchURL, kiwix::urlDecode(pattern));
   }
 
+  const std::string archiveUuid = getArchiveId(*archive);
```
Comment (mgautierfr) [inline src/server/internalServer.cpp:1064]: `getArchiveId` (and others) has been introduced by 8b12434 to avoid duplication between `kiwix::Reader::getId` and `kiwix::Book::update` which now use the `zim::Archive`.

Now that `kiwix::Reader` has been removed (and `getArchiveId` implementation simplified) you would tend to directly use `archive.getUuid()`

**Review by mgautierfr — CHANGES_REQUESTED** [review]
> !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
!!!      QUESTION: Do we need those resources?         !!!
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!


For favicon, I not sure but I think we should keep them.
Browsers may do some kind of auto discovery of the favicon by directly request them even if they are not declared in the html. @kelson42 may have a better idea here.

For `search-icon.svg`, it is referenced in `static/skin/index.css`
`search.svg` seems indeed not used. I think we can remove it.

**Comment by mgautierfr** [comment]
Real favicon generator may help us to know what we need to keep : https://realfavicongenerator.net/favicon_checker?protocol=http&site=dev.library.kiwix.org

File: src/server/internalServer.cpp (hunk 1 above)
Comment (veloman-yunkan) [reply src/server/internalServer.cpp]: Done

File: static/skin/index.css [hunk 2]
```diff
@@ -105,7 +105,7 @@ body {
     border-radius: 10px;
     border: solid 1px #b5b2b2;
     padding: 10px;
-    background-image: url('./search-icon.svg');
+    background-image: url('../skin/search-icon.svg?KIWIXCACHEID');
```
Comment (mgautierfr) [inline static/skin/index.css:108]: Why the change in the path (`../skin`) ?

**Review by mgautierfr — COMMENTED** [review]
I have one open question but else, it is a approval.

File: static/skin/index.css (hunk 2 above)
Comment (veloman-yunkan) [reply static/skin/index.css:108]: Because relative URLs to static resources are recognized by the `kiwix-resources` script using the `/skin/` anchor

**Comment by mgautierfr** [comment]
Please rebase-fixup on new master and we should be good to merge.

**Comment by veloman-yunkan** [comment]
> Please rebase-fixup on new master and we should be good to merge.

Done. The only change in addition to rebase-fixup is the edited commit message of 415ec4109989ef2d2b7443f3dbcd54081b934115 (the question answered by https://github.com/kiwix/libkiwix/pull/833#pullrequestreview-1139011544 was removed).

**Review by mgautierfr — APPROVED** [review]
